stonesoup/resampler/particle.py

```diff
@@ -23,25 +22,24 @@ def resample(self, particles):
             The resampled particles
         """
 
+        if particles and not isinstance(particles, Particles):
+            particles = Particles(particle_list=particles)
         n_particles = len(particles)
         weight = Probability(1/n_particles)
-        particles_sorted = sorted(particles, key=attrgetter('weight'), reverse=False)
-        cdf = np.cumsum([p.weight for p in particles_sorted])
+        if weight_order is None:
+            weight_order = np.argsort(particles.weight, kind='stable')
+        weight_order = np.array(weight_order)
+        cdf = [v.log_value for v in np.cumsum(particles.weight[weight_order])]
 
         # Pick random starting point
         u_i = np.random.uniform(0, 1 / n_particles)
-        new_particles = []
 
         # Cycle through the cumulative distribution and copy the particle
         # that pushed the score over the current value
-        for j in range(n_particles):
-
-            u_j = u_i + (1 / n_particles) * j
-
-            particle = particles_sorted[np.argmax(u_j < cdf)]
-            new_particles.append(
-                Particle(particle.state_vector,
-                         weight=weight,
-                         parent=particle))
-
+        u_j = u_i + (1 / n_particles) * np.arange(n_particles)
+        index = weight_order[np.searchsorted(cdf, np.log(u_j))]
+        new_particles = Particles(state_vector=particles.state_vector[:, index],
+                                  weight=[weight]*n_particles,
+                                  parent=Particles(state_vector=particles.state_vector[:, index],
+                                                   weight=[weight]*n_particles))
```

Comment:
This should be the parents weights.
```suggestion
                                                   weight=particles.weight[index]))
```